### PR TITLE
Add support for web3

### PR DESCRIPTION
**Pull request by levino**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme]
- [ ] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes). <----- **Not sure whether I really avoided them all.**
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration` <----- **Please see below**
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

Todos:

- [x] Improve file structure to reflect structure of web3
- [x] Add more test files
- [x] Add all collaborators
- [x] Make build pass (current errors see below)

### CODE REVIEW

**Comment by Levino** [comment]
I am still working on this, thus marking it work in progress. Will try finish within a couple of days.

**Comment by Levino** [comment]
Please take note that currently web3 provides its [own types](https://github.com/ethereum/web3.js/blob/c0e727eac7c16826c39fd43a03a4930b5ff6fff8/packages/web3/types.d.ts). As they are broken and of questionable quality I think it makes more sense to maintain them here. I opened an [issue](https://github.com/ethereum/web3.js/issues/1658) and [pr](https://github.com/ethereum/web3.js/pull/1660) to have them removed from the web3 codebase.

**Comment by Levino** [comment]
Meanwhile I get the feeling that spreading the definitions over multiple files like I did is an anti pattern. I am happy to merge all definitions into one flat file (again) even though this will be a long file and difficult to maintain. Please advise on what to do.

**Comment by eswarasai** [comment]
@Levino -- What all are pending for this PR to be approved?

The best pattern would be to create directories for each separate `typings` and keep them under that folder and import them in the main folder. For example, `web3-eth-abi`, `web3-eth-accounts` will have their respective typings and can be imported in `web3` `index.d.ts`. Makes sense?

**Comment by Levino** [comment]
@eswarasai Please improve / add your suggestions the codebase and make a PR to my branch.

**Comment by LogvinovLeon** [comment]
This PR contains a lot of types that are not web3-specific but rather ethereum-specific.
Puting them in this package will force other definitions to either import `web3` without actually needing it or just copying those definitions. I've created a package with ethereum-specific types. https://www.npmjs.com/package/ethereum-types And opened the PR to types-publisher to whitelist it https://github.com/Microsoft/types-publisher/pull/466 That way we'll be able to reuse all those types like `Unit`, `ABI`, `Transaction`, `Provider` in all packages. Not just web3.

**Comment by raulk** [comment]
Eager to see this merged!

**Comment by armanio123** [comment]
I agree with @LogvinovLeon that is good to factor out the ethereum protocol. Nevertheless, while we figure out the best approach I will merge this PR and things can be factored out later.

File: types/web3/providers.d.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
+interface JsonRPCRequest {
```
Comment (eswarasai) [inline types/web3/providers.d.ts]: @Levino -- Any harm in exporting the interfaces `JsonRPCRequest` and `JsonRPCResponse`? I'd like to have the typings for `payload` as well as `response` for this method -- 
```
web3.currentProvider.send(payload, (err, response) => {}))
```

Also, another thing I've noticed is that currently we don't have any interface defined for Event/Log Filters. Worth defining an interface called `FilterOptions` and use it across maybe? --
```
export interface FilterOptions {
    fromBlock?: BlockType;
    toBlock?: BlockType;
    address: string;
    topics?: Array<string | string[]>;
}
```

Let me know if you'd like me to create a PR for these two and update them. I'd be more than happy to. Thanks :)

**Comment by raulk** [comment]
We've been making improvements on these bindings in the EthQL repo: https://github.com/ConsenSys/ethql/tree/master/src/types/web3

Particularly around batching, utils, storage, etc.

File: types/web3/providers.d.ts (hunk 1 above)
Comment (Levino) [reply types/web3/providers.d.ts]: I just copied the existing type definitions I found at the the `web3` repos source code. Feel free to make a PR for the improvements you wish to have.

**Comment by Levino** [comment]
@raulk Please bring all your improvements here via PRs. Why would you maintain your own definitions elsewhere?

**Comment by raulk** [comment]
@levino I will batch submit a PR when I consider the typings to be stable and complete. Waiting on PRs to be merged upstream is not an option for me, hence the fork.

File: types/web3/providers.d.ts (hunk 1 above)
Comment (eswarasai) [reply types/web3/providers.d.ts]: @Levino -- Sure. Will soon create a PR for the above changes. Thanks!

**Pull request merged**